src/librustc_passes/ast_validation.rs

```diff
@@ -449,38 +471,27 @@ fn validate_generics_order<'a>(
 impl<'a> Visitor<'a> for AstValidator<'a> {
     fn visit_expr(&mut self, expr: &'a Expr) {
         match expr.node {
-            ExprKind::IfLet(_, ref expr, _, _) | ExprKind::WhileLet(_, ref expr, _, _) =>
-                self.while_if_let_ambiguity(&expr),
+            ExprKind::Let(_, _) if !self.is_let_allowed => {
+                self.err_handler()
+                    .struct_span_err(expr.span, "`let` expressions are not supported here")
+                    .note("only supported directly in conditions of `if`- and `while`-expressions")
+                    .note("as well as when nested within `&&` and parenthesis in those conditions")
+                    .emit();
+            }
+            _ if self.visit_expr_with_let_maybe_allowed(&expr) => {
```

Comment:
I find this setup to be somewhat confusing, but can't come up with a good solution either.

Reply:
An alternative is to fuse `visit_expr_with_let_maybe_allowed` with `visit_expr` but I think that would mix the problems more making it less self-contained.

Reply:
my worry is maintainability. There are quite some invariants to uphold now that weren't there before. Future changes may accidentally add new recursive calls to `visit_expr` without unsetting the `let` flag

Reply:
Sure, but that risk is inherent with a stateful visitor algorithm like the AST validation uses, and indeed, there have been bugs wrt. nested `impl Trait` as a result of that in the past. If you want to get rid of the risk entirely, you need to scrap the stateful visitor pattern approach in favor of a more functional approach (e.g. pass along state to recursive calls and in so doing you avoid effects on siblings).

As for forgetting to unset, it is always unset by default (at the end of the function) unless you explicitly return early, and the goal of setting things up with `visit_expr_with_let_maybe_allowed` was to make that more unlikely. As you see, I also tried to empty `visit_expr` as much as possible (added `obsolete_in_place`) to make the overall algorithm clearer. I also added a bunch of comments... :) Perhaps more would help?

Reply:
Hmm.... what do you think about wrapping `visit_expr`'s body in a closure before which the current flag is stored, set to `false` and restored afterwards? All the code that cares about the `let` flag can check the stored value (given as a closure argument) and has to manually wrap any calls it does with setting the `let` flag to `true`. This way only the code which actually cares about the flag has to take care to use it, and any bugs will be of the sort where you get false positives instead of having false negatives. The way it's setup right now one would silently introduce false negatives when doing something wrong.

Reply:
Implemented in https://github.com/rust-lang/rust/pull/60861/commits/f607ff1cd2356b99af2b8bdf461458fb027f4d5a.

Reply:
Awesome, thanks!

I think I already found the first example that (if const generics become more powerful), would have been wrongly accepted by the previous validation impl

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8104673337c46eddce70cc80bd76785b

Reply:
(I think both the new and old approach would have caught that as you must first go through an `ExprKind::Block(..)` which would have set `is_let_allowed = false` below the block in the tree. However, if the block was not allowed, I think both the new and the approach would have failed and you would need to set `is_let_allowed = false` when entering `visit_ty`.)